### PR TITLE
[DO NOT MERGE YET] DGX Spark integrations

### PR DESCRIPTION
**Pull request by rltakashige**
## Motivation

TODO BEFORE MERGE:
- test for no GLM regression.
- `nix run exo` to just work. Or at least `nix run exo-cuda`

Part 1 of getting EXO to run well on the DGX Spark and other Linux CUDA machines outside of private betas.
`nix run .#exo-cuda`
or `just sync-cuda && uv run exo`

## Changes
Cache a nix build of the dependencies
Add setup scripts
Make Linux look nice on the dashboard

## Caveats
The connectivity between a Mac and Spark is still super dodgy.

Screenshot 2026-03-11 at 18 33 23

### CODE REVIEW

File: src/exo/shared/types/worker/instances.py [hunk 1]
```diff
@@ -35,8 +36,11 @@ class MlxJacclInstance(BaseInstance):
     jaccl_coordinators: dict[NodeId, str]
 
 
-# TODO: Single node instance
```
Comment (Evanev7) [inline src/exo/shared/types/worker/instances.py:38]: :'(

File: src/exo/master/api.py [hunk 2]
```diff
@@ -220,6 +220,14 @@ def __init__(
         self.paused: bool = False
         self.paused_ev: anyio.Event = anyio.Event()
 
+        try:
+            import vllm
+
+            logger.info(f"Found vllm version {vllm.__version__}")
+            self._vllm_available = True
```
Comment (Evanev7) [inline src/exo/master/api.py]: should be in state proper.

File: src/exo/worker/engines/mlx/generator/batch_generate.py [hunk 3]
```diff
@@ -291,14 +295,8 @@ def step(self) -> list[tuple[int, GenerationResponse]]:
...
-                try:
-                    mlx_stats = self._exo_gen.stats()
-                except ZeroDivisionError:
-                    mlx_stats = None
                 stats = GenerationStats(
-                    prompt_tps=float(mlx_stats.prompt_tps)
-                    if mlx_stats is not None and mlx_stats.prompt_time > 0
-                    else 0.0,
+                    prompt_tps=state.prefill_tps,
                     generation_tps=float(generation_tps),
```
Comment (Evanev7) [inline src/exo/worker/engines/mlx/generator/batch_generate.py]: do the thing!!!

File: src/exo/vllm_patches/growable_cache.py
Comment (Evanev7) [inline src/exo/vllm_patches/growable_cache.py:1]: kill die death

File: src/exo/worker/engines/vllm/prompt_format.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+from exo.worker.engines.mlx.utils_mlx import (
+    apply_chat_template,
+    get_eos_token_ids_for_model,
+)
+
+
+def format_vllm_prompt(
+    engine: LLMEngine, params: TextGenerationTaskParams
+) -> tuple[list[int], str, int]:
+    tokenizer = TokenizerWrapper(engine.get_tokenizer())
```
Comment (Evanev7) [inline src/exo/worker/engines/vllm/prompt_format.py:18]: we should have our own wrapper so we can just use the tokenizers Tokenizer

File: src/exo/worker/engines/vllm/vllm_generator.py [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+)
+from exo.worker.engines.vllm.prompt_format import (
+    format_vllm_prompt,
+    make_vllm_sampling_params,
+)
+from exo.worker.runner.bootstrap import logger
+from exo.worker.runner.llm_inference.tool_parsers import ToolParser, infer_tool_parser
+
+
+def _build_layer_groups(kv_cache_config: object) -> list[int]:
```
Comment (Evanev7) [inline src/exo/worker/engines/vllm/vllm_generator.py]: type stubs?

File: src/exo/worker/engines/vllm/growable_cache.py [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+    original = Worker.initialize_from_config
+
+    def patched(self: "Worker", kv_cache_config: "object") -> None:
+        original(self, kv_cache_config)
+        _growable_model_runner_ref[0] = self.model_runner
+
+    Worker.initialize_from_config = patched  # type: ignore
+
+
+_growable_model_runner_ref: list["object | None"] = [None]
```
Comment (Evanev7) [inline src/exo/worker/engines/vllm/growable_cache.py]: put at top

File: src/exo/worker/engines/vllm/vllm_generator.py [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+            torch.distributed.destroy_process_group()
+
+
+_weight_loading_callback: Callable[[int, int], None] | None = None
+_weight_loading_patched = False
+
+_LAYER_INDEX_PATTERN = re.compile(r"\.layers\.(\d+)\.")
+
+
+def _get_total_layers(model_dir: Path) -> int:
```
Comment (Evanev7) [inline src/exo/worker/engines/vllm/vllm_generator.py]: use the model card

File: src/exo/worker/engines/vllm/vllm_generator.py [hunk 8]
```diff
@@ -0,0 +1,583 @@
...
+    for tensor_spec in kv_cache_config.kv_cache_tensors:  # type: ignore
+        for name in tensor_spec.shared_by:  # type: ignore
+            layer_to_group.append(group_lookup[name])  # type: ignore
+    return layer_to_group
+
+
+@dataclass
+class _EngineRequest:
+    uid: int
+    request_id: str
```
Comment (Evanev7) [inline src/exo/worker/engines/vllm/vllm_generator.py:58]: can merge request_id's with uids

File: src/exo/worker/engines/vllm/vllm_generator.py [hunk 9]
```diff
@@ -0,0 +1,583 @@
...
+
+    engine_args = EngineArgs(
+        model=model_path,
+        served_model_name=str(model_id),
+        gpu_memory_utilization=0.05,
+        trust_remote_code=trust_remote_code,
+        load_format="fastsafetensors",
+        enable_prefix_caching=False,
+        attention_backend="TRITON_ATTN",
+        enforce_eager=True,
```
Comment (Evanev7) [inline src/exo/worker/engines/vllm/vllm_generator.py:575]: maybe default

File: src/exo/worker/engines/vllm/growable_cache.py [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+import torch
+
+from exo.shared.logging import logger
+
+INITIAL_FRACTION = 0.05
+GROWTH_HEADROOM_BYTES = 512 * 1024 * 1024
+MIN_GROWTH_BLOCKS = 16
+
+_patched = False
+_exo_prefix_cache_ref: list["object | None"] = [None]
```
Comment (Evanev7) [inline src/exo/worker/engines/vllm/growable_cache.py]: add getter & setter functions via global

File: src/exo/worker/engines/vllm/vllm_generator.py [hunk 11]
```diff
@@ -0,0 +1,583 @@
...
+                setattr(mod, attr, _noop_metadata)
+
+
+def load_vllm_engine(
+    model_path: str,
+    model_id: ModelId,
+    trust_remote_code: bool,
+    on_layer_loaded: Callable[[int, int], None] | None = None,
+) -> tuple[LLMEngine, ToolParser | None, KVPrefixCache]:
+    global _weight_loading_callback
```
Comment (Evanev7) [inline src/exo/worker/engines/vllm/vllm_generator.py]: global feels unnecessary

File: src/exo/worker/engines/kv_cache.py [hunk 12]
```diff
@@ -0,0 +1,310 @@
+# pyright: reportUnknownMemberType=false, reportAttributeAccessIssue=false
+# pyright: reportUnknownVariableType=false, reportUnknownParameterType=false
+# pyright: reportUnknownArgumentType=false
```
Comment (Evanev7) [inline src/exo/worker/engines/kv_cache.py]: dont type ignore

File: src/exo/worker/engines/vllm/kv_cache.py
Comment (Evanev7) [inline src/exo/worker/engines/vllm/kv_cache.py:1]: seems like a vllm specific cache.

File: src/exo/worker/runner/llm_inference/batch_generator.py [hunk 13]
```diff
@@ -1,9 +1,12 @@
+from __future__ import annotations
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/batch_generator.py]: personally i avoid __future__ annotations if possible

File: src/exo/worker/runner/llm_inference/batch_generator.py [hunk 14]
```diff
@@ -304,7 +310,7 @@ def close(self) -> None:
 
 @dataclass(eq=False)
 class BatchGenerator(InferenceGenerator):
-    model: Model
+    model: Model | None
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/batch_generator.py]: the batch generator should not own the model if possible- what blockers are there to this?

File: src/exo/worker/runner/llm_inference/batch_generator.py [hunk 15]
```diff
@@ -330,21 +337,14 @@ class BatchGenerator(InferenceGenerator):
...
-        )
-
-    def warmup(self):
-        self.check_for_cancel_every = warmup_inference(
-            model=self.model,
-            tokenizer=self.tokenizer,
-            group=self.group,
-            model_id=self.model_id,
-        )
+    def warmup(self) -> None:
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/batch_generator.py:327]: should defer to engine

File: src/exo/worker/runner/llm_inference/runner.py [hunk 16]
```diff
@@ -247,12 +278,11 @@ def shutdown(self, task: Task):
         self.acknowledge_task(task)
         if isinstance(self.generator, InferenceGenerator):
             self.generator.close()
-        mx.clear_cache()
-        import gc
-
+        else:
+            self.generator.shutdown_cleanup()
         gc.collect()
-        self.send_task_status(task.task_id, TaskStatus.Complete)
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/runner.py:254]: check it

File: src/exo/worker/runner/llm_inference/runner.py [hunk 17]
```diff
@@ -247,12 +278,11 @@ def shutdown(self, task: Task):
         self.acknowledge_task(task)
         if isinstance(self.generator, InferenceGenerator):
             self.generator.close()
-        mx.clear_cache()
-        import gc
-
+        else:
+            self.generator.shutdown_cleanup()
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/runner.py]: check it

File: src/exo/worker/runner/llm_inference/runner.py [hunk 18]
```diff
@@ -154,22 +182,28 @@ def handle_first_task(self, task: Task):
...
                 self.send_task_status(task.task_id, TaskStatus.Complete)
                 self.update_status(RunnerConnected())
                 logger.info("runner connected")
 
             # we load the model if it's connected with a group, or idle without a group. we should never tell a model to connect if it doesn't need to
-            case LoadModel() if isinstance(self.generator, Builder) and (
+            case LoadModel() if (
                 (
-                    isinstance(self.current_status, RunnerConnected)
+                    isinstance(self.generator, MlxBuilder)
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/runner.py:196]: can just be isinstance(Builder)

File: src/exo/worker/runner/llm_inference/runner.py [hunk 19]
```diff
@@ -429,4 +481,77 @@ def build(
             device_rank=device_rank,
             cancel_receiver=self.cancel_receiver,
             event_sender=self.event_sender,
+            _gen=gen,
         )
+
+    def shutdown_cleanup(self) -> None:
+        mx.clear_cache()
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/runner.py]: move to inner generator

File: src/exo/worker/runner/bootstrap.py [hunk 20]
```diff
@@ -1,16 +1,47 @@
...
+        if not driver.exists():
+            continue
+        for lib_name in _CUDA_HOST_LIBS:
+            lib_path = search_dir / lib_name
+            if lib_path.exists():
+                try:
+                    ctypes.CDLL(str(lib_path), mode=ctypes.RTLD_GLOBAL)
+                    logger.info(f"Loaded CUDA host lib: {lib_path}")
+                except OSError as e:
+                    logger.warning(f"Failed to load {lib_path}: {e}")
```
Comment (Evanev7) [inline src/exo/worker/runner/bootstrap.py]: \+ raise

File: src/exo/worker/runner/llm_inference/runner.py (hunk 18 above)
Comment (rltakashige) [reply src/exo/worker/runner/llm_inference/runner.py:196]: I do want the generator to explicitly be an mlx builder if there is an mlx group.

File: src/exo/worker/runner/llm_inference/runner.py (hunk 19 above)
Comment (rltakashige) [reply src/exo/worker/runner/llm_inference/runner.py]: Not sure I understand why we'd do this?

File: src/exo/worker/runner/llm_inference/batch_generator.py (hunk 14 above)
Comment (rltakashige) [reply src/exo/worker/runner/llm_inference/batch_generator.py]: Made an attempt at this by moving generator creation before the warmup phase.

File: src/exo/worker/engines/vllm/prompt_format.py (hunk 4 above)
Comment (rltakashige) [reply src/exo/worker/engines/vllm/prompt_format.py:18]: Probably a TODO for a separate PR. Added it as a comment.
